execution/src/windows/execution.cc

```diff
@@ -106,16 +110,24 @@ namespace leatherman { namespace execution {
     static tuple<scoped_handle, scoped_handle> CreatePipeThrow()
     {
         static LONG counter = 0;
+        static boost::uuids::random_generator rand_uuid;
 
         SECURITY_ATTRIBUTES attributes = {};
         attributes.nLength = sizeof(SECURITY_ATTRIBUTES);
         attributes.bInheritHandle = TRUE;
         attributes.lpSecurityDescriptor = NULL;
 
-        // Format a name for the pipe based on the process and counter
-        wstring name = boost::nowide::widen((boost::format("\\\\.\\Pipe\\leatherman.%1%.%2%") %
+        // Format a name for the pipe. Use the thread id to ensure no two threads try to use the same
+        // pipe, and a counter to generate multiple pipes for the same process invocation.
+        // A scenario exists using timeouts where we could release the invoking end of a named pipe
+        // but the other end doesn't release. Then the invoking thread shuts down and another with
+        // the same thread id is started and reconnects to the existing named pipe. Use the process
+        // id and a random UUID to make that highly unlikely.
+        wstring name = boost::nowide::widen((boost::format("\\\\.\\Pipe\\leatherman.%1%.%2%.%3%.%4%") %
+            GetCurrentProcessId() %
             GetCurrentThreadId() %
-            InterlockedIncrement(&counter)).str());
+            InterlockedIncrement(&counter) %
```

Comment:
Statistically unlikely. We do know we'll open several unique pipes for the same command, however.